src/data/polynomial/unit_trinomial.lean

```diff
@@ -0,0 +1,315 @@
+  if_neg hkm.ne', if_pos rfl, if_neg hmn.ne, zero_add, add_zero]
+
+lemma trinomial_trailing_coeff' (hkm : k < m) (hmn : m < n) :
+  (C u * X ^ k + C v * X ^ m + C w * X ^ n).coeff k = u :=
+by rw [coeff_add, coeff_add, coeff_C_mul_X_pow, coeff_C_mul_X_pow, coeff_C_mul_X_pow,
+  if_pos rfl, if_neg hkm.ne, if_neg (hkm.trans hmn).ne, add_zero, add_zero]
+
+lemma trinomial_nat_degree (hkm : k < m) (hmn : m < n) (hw : w ≠ 0) :
+  (C u * X ^ k + C v * X ^ m + C w * X ^ n).nat_degree = n :=
+begin
+  refine nat_degree_eq_of_degree_eq_some (le_antisymm (sup_le (λ i h, _))
+    (le_degree_of_ne_zero (by rwa trinomial_leading_coeff' hkm hmn))),
+  replace h := support_trinomial' k m n u v w h,
+  rw [mem_insert, mem_insert, mem_singleton] at h,
+  rcases h with rfl | rfl | rfl,
+  { exact with_bot.coe_le_coe.mpr (hkm.trans hmn).le },
+  { exact with_bot.coe_le_coe.mpr hmn.le },
+  { exact le_rfl },
+end
+
+lemma trinomial_nat_trailing_degree (hkm : k < m) (hmn : m < n) (hu : u ≠ 0) :
+  (C u * X ^ k + C v * X ^ m + C w * X ^ n).nat_trailing_degree = k :=
+begin
+  refine nat_trailing_degree_eq_of_trailing_degree_eq_some (le_antisymm (le_inf (λ i h, _))
+    (le_trailing_degree_of_ne_zero (by rwa trinomial_trailing_coeff' hkm hmn))).symm,
+  replace h := support_trinomial' k m n u v w h,
+  rw [mem_insert, mem_insert, mem_singleton] at h,
+  rcases h with rfl | rfl | rfl,
+  { exact le_rfl },
+  { exact with_top.coe_le_coe.mpr hkm.le },
+  { exact with_top.coe_le_coe.mpr (hkm.trans hmn).le },
+end
+
+lemma trinomial_leading_coeff (hkm : k < m) (hmn : m < n) (hw : w ≠ 0) :
+  (C u * X ^ k + C v * X ^ m + C w * X ^ n).leading_coeff = w :=
+by rw [leading_coeff, trinomial_nat_degree hkm hmn hw, trinomial_leading_coeff' hkm hmn]
+
+lemma trinomial_trailing_coeff (hkm : k < m) (hmn : m < n) (hu : u ≠ 0) :
+  (C u * X ^ k + C v * X ^ m + C w * X ^ n).trailing_coeff = u :=
+by rw [trailing_coeff, trinomial_nat_trailing_degree hkm hmn hu, trinomial_trailing_coeff' hkm hmn]
+
+lemma trinomial_mirror (hkm : k < m) (hmn : m < n) (hu : u ≠ 0) (hw : w ≠ 0) :
+  (C u * X ^ k + C v * X ^ m + C w * X ^ n).mirror =
+    C w * X ^ k + C v * X ^ (n - m + k) + C u * X ^ n :=
+by rw [mirror, trinomial_nat_trailing_degree hkm hmn hu, reverse, trinomial_nat_degree hkm hmn hw,
+  reflect_add, reflect_add, reflect_C_mul_X_pow, reflect_C_mul_X_pow, reflect_C_mul_X_pow,
+  rev_at_le (hkm.trans hmn).le, rev_at_le hmn.le, rev_at_le le_rfl, add_mul, add_mul, mul_assoc,
+  mul_assoc, mul_assoc, ←pow_add, ←pow_add, ←pow_add, nat.sub_add_cancel (hkm.trans hmn).le,
+  nat.sub_self, zero_add, add_comm, add_assoc, add_comm (C u * X ^ n)]
+
+end semiring
+
+variables (p q : ℤ[X])
+
+/-- A unit trinomial is a trinomial with unit coefficients. -/
+def is_unit_trinomial := ∃ {k m n : ℕ} (hkm : k < m) (hmn : m < n) {u v w : ℤ}
+  (hu : is_unit u) (hv : is_unit v) (hw : is_unit w), p = C u * X ^ k + C v * X ^ m + C w * X ^ n
```

Comment:
In my experience, working with `units` is nicer than `is_unit`, so:
```suggestion
def is_unit_trinomial := ∃ {k m n : ℕ} (hkm : k < m) (hmn : m < n) {u v w : units ℤ}
  p = C u * X ^ k + C v * X ^ m + C w * X ^ n
```

Reply:
Sure. It broke a few `rcases with rfl`, but didn't cause too much trouble.